### ISSUE TITLE
SIMD CRC crates

### ISSUE THREAD

**Issue by srijs** [alexcrichton/flate2-rs#173]
Hi!

This is funny, I just saw that you published `flate2-crc` a couple of hours ago.

Coincidentally, I wrote and published [`crc32fast`](https://crates.io/crates/crc32fast) two days ago based on the same Intel paper, in order to speed up the 0.5.0 release of the `zip` crate.

Just wanted to raise some awareness and see if it makes sense to join forces somehow!

PS: If I get some time, I'll try and benchmark our two approaches to see what they could learn from each other :)

**Comment by srijs** [comment]
FWIW, performance seems to be largely on-par, which makes me wonder if some of the unrolling optimizations in `crc32fast` are really necessary:

```
test crc32fast_1000000  ... bench:      78,839 ns/iter (+/- 16,440) = 12684 MB/s
test crc32fast_65000    ... bench:       5,221 ns/iter (+/- 1,871) = 12449 MB/s
test crc32fast_8        ... bench:          70 ns/iter (+/- 23) = 114 MB/s
test flate2_crc_1000000 ... bench:      82,647 ns/iter (+/- 38,070) = 12099 MB/s
test flate2_crc_65000   ... bench:       6,136 ns/iter (+/- 2,466) = 10593 MB/s
test flate2_crc_8       ... bench:          29 ns/iter (+/- 22) = 275 MB/s
```

```
test crc32fast_1000000  ... bench:      81,187 ns/iter (+/- 16,621) = 12317 MB/s
test crc32fast_65000    ... bench:       5,146 ns/iter (+/- 1,521) = 12631 MB/s
test crc32fast_8        ... bench:          64 ns/iter (+/- 35) = 125 MB/s
test flate2_crc_1000000 ... bench:      77,260 ns/iter (+/- 18,388) = 12943 MB/s
test flate2_crc_65000   ... bench:       4,913 ns/iter (+/- 746) = 13230 MB/s
test flate2_crc_8       ... bench:          18 ns/iter (+/- 9) = 444 MB/s
```

**Comment by oyvindln** [comment]
Alternatively, make a SIMD implementation PR for the [crc](https://crates.io/crates/crc) crate since it's already use by a fair number of other crates.

**Comment by alexcrichton** [comment]
Oh dear, sorry for the duplicating! I spent a full day scouring crates.io and trying to modify existing crates and came up empty! It's good that the perf is on-par though :)

I'd be fine switching over, although it looks like the internal `State` holds a lot of crc registers, but is that necessary? I was also very wary of diverging much from the original paper myself so I tried to make sure everything followed as closely as possible and was documented where it differed, but I unfortunately found it somewhat difficult to follow your implementation :( 

@oyvindln I personally opted to not modify the `crc` crate because I wasn't really sure which CRC algorithm it was implementing (apparently there are tons) and I'd already made the wrong choice there before (tried out using a crc crate that already had simd support but turned out to be the wrong crc). I also ideally didn't want users of `flate2` to pull in support for lots of crc implementations they don't end up needing

**Comment by srijs** [comment]
The reason for holding multiple internal CRC registers in the state is that the implementation follows the "Parallel Folding of 128-bit Data chunks" recommendation in the paper (aiming to perform up to 4 folds in parallel) and defers the final reduction step until all data has been processed.

On machines with wider SIMD lanes this should theoretically give it an edge, but I also like the (relative) simplicity of your implementation! Given we both use MIT/Apache, we should be able to move code around if it makes sense to us.

I also looked into modifying the `crc` crate, but opted against that because I only needed a fast CRC32 computation using the IEEE polynomials, so I wasn't sure how that would fit into the generic framework that this crate offers.

I'll try to get a better perf comparison on a larger machine to see if my more complex implementation is worth it!

**Comment by srijs** [comment]
Finally got the chance to run a few benchmark on larger machines, and at this point it looks like the optimizations described in the Intel paper (or at least as I ended up implementing them) are not worth the complexity, as performance continues to be largely on par with the simple one-fold-at-a-time implementation.

I'm keen to hear your thoughts on this @alexcrichton, but I'm inclined at this point to either adapt your simd code for the `crc32fast` crate (with the appropriate attribution of course), and let you decide whether you'd like to keep `flate2-crc` or switch over to `crc32fast` (I'm also happy to give you publish rights if that helps de-risk things for you).

**Comment by alexcrichton** [comment]
@srijs ok thanks for the investigation! Out of curiosity, I wasn't sure why the four-at-a-time fold was recommended in the intel paper, but is that not necessary either basically? 

I'm totally down for depending on a different crate, so I'd be ok relying on `crc32fast` for this!

**Comment by srijs** [comment]
Just published version 1.0.4 of `crc32fast` which includes your simplified SIMD implementation!
